Student/Widgets/AnnouncementsWidget/Controller/AnnouncementsProvider.swift

```diff
@@ -19,46 +19,81 @@
 import Core
 import WidgetKit
 
-class AnnouncementsProvider: CommonWidgetProvider<AnnouncementsEntry> {
+class AnnouncementsProvider: TimelineProvider {
+    typealias Entry = AnnouncementsEntry
+    private let env = AppEnvironment.shared
     private var colors: Store<GetCustomColors>?
     private var courses: Store<GetAllCourses>?
     private var announcements: Store<GetWidgetAnnouncements>?
 
-    init() {
-        super.init(loggedOutModel: AnnouncementsEntry(isLoggedIn: false), timeout: GetWidgetAnnouncements.Timeout)
+    // MARK: - TimelineProvider Protocol
+
+    func placeholder(in context: TimelineProvider.Context) -> Entry { .publicPreview }
+
+    func getSnapshot(in context: TimelineProvider.Context, completion: @escaping (AnnouncementsEntry) -> Void) {
+        completion(placeholder(in: context))
+    }
+
+    func getTimeline(in context: TimelineProvider.Context, completion: @escaping @Sendable (Timeline<AnnouncementsEntry>) -> Void) {
+
+        if context.isPreview {
+            let timeline = Timeline(entries: [placeholder(in: context)], policy: .never)
+            completion(timeline)
+            return
+        }
+
+        guard let session = LoginSession.mostRecent else {
+            let refreshDate = Clock.now.addingTimeInterval(.widgetRefresh)
+            completion(Timeline(entries: [.loggedOutModel], policy: .after(refreshDate)))
+            return
+        }
+
+        setupEnvironment(with: session)
+        fetchColors(completion)
     }
 
-    override func fetchData() {
+    // MARK: - Private
+
+    private func setupEnvironment(with session: LoginSession) {
+        env.app = .student
+        guard let session = LoginSession.mostRecent else {
+            env.widgetUserDidLogout()
+            return
+        }
+        if let current = env.currentSession, current == session { return }
+        env.userDidLogin(session: session, isSilent: true)
+    }
+
+    func fetchColors(_ completion: @escaping @Sendable (Timeline<AnnouncementsEntry>) -> Void) {
         colors = env.subscribe(GetCustomColors())
         colors?.refresh { [weak self] _ in
-            guard let self = self, let colors = self.colors, !colors.pending else { return }
-            self.fetchCourses()
+            self?.fetchCourses(completion)
         }
     }
 
-    private func fetchCourses() {
+    func fetchCourses(_ completion: @escaping @Sendable (Timeline<AnnouncementsEntry>) -> Void) {
+        guard let colors = self.colors, !colors.pending else { return }
```

Comment:
```suggestion
        guard let colors, !colors.pending else { return }
        
```
(and please add a newline)

Please do the same in the other methods below, as well.

Reply:
Done.